File: content/problems/0034-find-first-and-last-position-of-element-in-sorted-array/solutions/two-lower-bounds/solution.cpp
class Solution {
    int lowerBound(const vector<int>& nums, long long target) {
        int left = 0, right = nums.size();
        while (left < right) { int middle = left + (right - left) / 2; if (nums[middle] < target) left = middle + 1; else right = middle; }
        return left;
    }
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        int first = lowerBound(nums, target);
        if (first == (int)nums.size() || nums[first] != target) return {-1, -1};
        return {first, lowerBound(nums, (long long)target + 1) - 1};
    }
};

File: content/problems/0034-find-first-and-last-position-of-element-in-sorted-array/solutions/two-lower-bounds/Solution.kt
class Solution {
    fun searchRange(nums: IntArray, target: Int): IntArray {
        val first = lowerBound(nums, target.toLong())
        if (first == nums.size || nums[first] != target) return intArrayOf(-1, -1)
        return intArrayOf(first, lowerBound(nums, target.toLong() + 1) - 1)
    }
    private fun lowerBound(nums: IntArray, target: Long): Int {
        var left = 0
        var right = nums.size
        while (left < right) { val middle = left + (right - left) / 2; if (nums[middle].toLong() < target) left = middle + 1 else right = middle }
        return left
    }
}
